chore: add testing example for local build (#579)
canonical/netplan: README.md

# Build using Meson

> A Makefile wrapper is also provided for simplified usage. For that approach, please refer to the [Build using Makefile][makefile-section] section below.

Reviewer: Should this be:
```suggestion
A Makefile wrapper is also provided for simplified usage. For that approach, please refer to the [Build using Makefile](#build-using-makefile) section below.
```

* meson compile -C build
* meson test -C build --verbose [TEST_NAME]
* meson install -C build --destdir ../tmproot

# Build using Makefile

Convenience targets are available via `make`:

- `make` or `make default`  
  Set up the build directory (`_build`) and build the project

- `make check`  
  Build and run all tests in `_build`

- `make linting`  
  Run Meson `linting` and `codestyle` test targets

- `make check-coverage`  
  Set up a coverage build directory (`_build-cov`), build, and run coverage tests

- `make install [DESTDIR=../tmproot]`  
  Build and install into a staging root (defaults to `../tmproot`)

- `make clean`  
  Remove generated build and test artifacts

- `make run ARGS='<command>'`  
  Run the locally built CLI with the appropriate environment, e.g.:
  ```sh
  make run ARGS="help"
  ```

# Test local build

After running:

```sh
make
make install
```

you can test the locally built `netplan` without installing it system-wide:

```sh
make run ARGS="<command>"
```

This wrapper sets the required environment variables automatically:

`NETPLAN_GENERATE_PATH`
`NETPLAN_CONFIGURE_PATH`
`LD_LIBRARY_PATH`
`PYTHONPATH`

> These are needed because the Python CLI resolves binary and library paths at runtime.

Reviewer: Let's not list them all, otherwise we have to update this list if they change. How about shortening to:
```suggestion
This wrapper sets the required environment variables (such as `NETPLAN_GENERATE_PATH`) automatically. These are needed because the Python CLI resolves binary and library paths at runtime.
```

**Example**:

```sh
# build and install
make
make install

# run a command (netplan info)
make run ARGS="info"
```

*output*:

```sh
netplan.io:
  website: "https://netplan.io/"
  features:
  - dhcp-use-domains
  - auth-phase2
  - ipv6-mtu
  - modems
  - sriov
  - openvswitch
  - activation-mode
  - eswitch-mode
  - infiniband
  - regdom
  - vrf
  - vxlan
  - virtual-ethernet
  - dbus-config
  - generated-suppl
```

> 

Reviewer: Let's tighten this up. We've already mentioned `make` and `make install` above, so we can drop that. Also, let's not include the full output, just the first few lines and `...`. Let's do something like this:
```suggestion
As an example, let's use `make run` to run `netplan info`:

```sh
$ make run ARGS='info'
netplan.io:
  website: "https://netplan.io/"
  features:
  - dhcp-use-domains
  - auth-phase2
  ...
```

```

# Build using Meson

A Makefile wrapper is also provided for simplified usage. For that approach, please refer to the [Build using Makefile][makefile-section] section below.

Steps to build Netplan using the [Meson](https://mesonbuild.com) build system inside the `build/` directory:

* meson setup build --prefix=/usr [-Db_coverage=true]
* meson compile -C build
* meson test -C build --verbose [TEST_NAME]
* meson install -C build --destdir ../tmproot

# Build using Makefile

Convenience targets are available via `make`:

- `make` or `make default`  
  Set up the build directory (`_build`) and build the project

- `make check`  
  Build and run all tests in `_build`

- `make linting`  
  Run Meson `linting` and `codestyle` test targets

> - `make check-coverage`  

Reviewer: I think we should drop this one (check-coverage), as it doesn't seem generally useful enough to include here.

  Set up a coverage build directory (`_build-cov`), build, and run coverage tests

- `make install [DESTDIR=../tmproot]`  
  Build and install into a staging root (defaults to `../tmproot`)

- `make clean`  
  Remove generated build and test artifacts

- `make run ARGS='<command>'`  

>   Run the locally built CLI with the appropriate environment, e.g.:

Reviewer: Nit, let's change to:

```suggestion
  Run the locally built netplan CLI with the appropriate environment, for example, to run `netplan help`:
```

* meson test -C build --verbose [TEST_NAME]
* meson install -C build --destdir ../tmproot

> # Test a local build (backend: networkd)

Reviewer: Seeing there's a nice Makefile that wraps Meson with the correct options, I think we should probably lean into that and update the section above to use the Makefile commands. It looks like the default DESTDIR is `../tmproot`, too.

For example, the first two (meson setup and meson compile) are covered by the default make target, so just typing `make`. Then test is `make check` and install is `make install`.

Reviewer: There's also a lot of build dependencies that need to be installed. I had to install these (I don't think I missed any) in addition to whatever was already on my system:

```
sudo apt install \
	meson \
	libyaml-dev \
	libsystemd-dev \
	pyflakes3 \
	pycodestyle \
	python3-coverage \
	pandoc \
	python3-cffi \
	libcmocka-dev \
	python3-dev
```

I think we should mention these steps too, from checkout to build.

Not for this PR, but soon I think we should switch to using `pyproject.toml` and `uv` to make development a lot easier. And/or using devcontainers to avoid messing so much with the local environment.

After `meson install -C _build --destdir ../tmproot`, test the local build without touching the system install by passing environment variables that redirect the CLI to your build tree (`NETPLAN_GENERATE_PATH`, `NETPLAN_CONFIGURE_PATH`, `LD_LIBRARY_PATH`, `PYTHONPATH`). These are needed because the Python CLI resolves binary and library paths at runtime.

> Prepare a test config (replace `eth0` with an interface present on your system):

Reviewer: I think instead of using a modifying command like `netplan try`, we should just drop this test config and show `netplan info` or something simpler as the example.

We could add a `make run` target which allowed ARGS to be passed, like `make run ARGS='info'` would be like `netplan info`. That could set all the environment variables.

Or if that's too ugly we could add a `netplan` script at the top level which sets the env vars and then calls `tmproot/usr/sbin/netplan`.

Email the list at [netplan-developers@lists.launchpad.net](mailto:netplan-developers@lists.launchpad.net).

> [makefile-section]: #build-using-makefile

Reviewer: Oh, I see what you've done here. Let's just put the link inline.
